### PR TITLE
Express endpoints for admin authentication

### PR DESCRIPTION
**Pull request by draycfarley**
### Summary 

This pull request implements an express endpoint for checking if a token is for an admin user or not.

### Test Plan 
Tested through Postman with my own Google account token as well as through adding unit tests which mock the Google client's response in Auth.test.ts.

image

### Notes 

We may want to change the test file to mock the Google client object instead of the getVerificationTicket in the future. I would like others' thoughts on this as well.

### CODE REVIEW

File: server/endpoints/Auth.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+/**
+   * Returns true if [netid] matches the netid in the email of the JSON
+   * web token. False otherwise.
+   * This method authenticates the user token through the Google API.
+   * @param token: google auth token
+   * @param netid: netid to verify
+   * @requires that you have a handleVerifyError, like as follows:
+   * verify(token, function(){//do whatever}).catch(function(error){
+   * handleVerifyError(error, res);
+   */
```
Comment (julian-londono) [inline server/endpoints/Auth.ts:22]: I don't think this function actually requires this anymore - seems like old documentation

File: server/endpoints/Auth.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  try {
+    if (token === null) {
+      console.log("Token was undefined in getVerificationTicket");
+      return null; // Token was undefined
+    }
+    const ticket = await client.verifyIdToken({
+      idToken: token,
+      audience: "836283700372-msku5vqaolmgvh3q1nvcqm3d6cgiu0v1.apps.googleusercontent.com", // Specify the CLIENT_ID of the app that accesses the backend
+      // Or, if multiple clients access the backend:
+      // [CLIENT_ID_1, CLIENT_ID_2, CLIENT_ID_3]
```
Comment (julian-londono) [inline server/endpoints/Auth.ts:33]: These comments about `CLIENT_ID` stuff is old boilerplate comments from a tutorial -  can be removed

File: server/endpoints/Auth.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+   * Used in the .catch when verify is used, handles whatever should be done
+   * @param errorObj (required) the error that is returned from the .catch
+   * @param res the response object
+   * @return {boolean} true if their token is too old, false if some other error
+   * @requires that you have the verify function, like as follows:
+   * verify(token, function(){//do whatever}).catch(function(error){
+   *        handleVerifyError(error, res);
+   * }
+   */
+export const handleVerifyError = (errorObj, res) => {
```
Comment (julian-londono) [inline server/endpoints/Auth.ts:54]: Related to comment above - I don't think `handleVerifyError` is ever explicitly used anywhere.  Could somehow be implicitly called but if not it should be removed

File: server/endpoints/Auth.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  }
+  return false;
+};
+
+// Get a user with this netId from the Users collection in the local database
+export const getUserByNetId = async (netId: string) => {
+  try {
+    const regex = new RegExp(/^(?=.*[A-Z0-9])/i);
+    if (regex.test(netId)) {
+      return await Students.findOne({ netId }).exec();
```
Comment (julian-londono) [inline server/endpoints/Auth.ts:72]: Not 100% positive our `_id` field in `Student` and the following are equivalent but the [Mongoose docs seem to recommend using `findById()`](https://mongoosejs.com/docs/api.html#model_Model.findOne) in this case

**Review by julian-londono — CHANGES_REQUESTED** [review]

**Comment by julian-londono** [comment]
### re: Thoughts on mocking the Google client object

I think this would be useful in the future as it would therefore add tests to `getVerificationTicket` itself.  For how, however, I think the tests set up so far are good.  I'd save that for a future PR.

File: server/endpoints/Auth.ts (hunk 4 above)
Comment (draycfarley) [reply server/endpoints/Auth.ts:72]: The netID is different from the document ID

**Comment by draycfarley** [comment]
@julian-londono Could you mark the changes as resolved?

**Review by julian-londono — APPROVED** [review]

**Pull request merged**